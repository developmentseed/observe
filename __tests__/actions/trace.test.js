/* global jest, it, expect, describe, fetch */

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import {
  startTrace,
  endTrace,
  pauseTrace,
  unpauseTrace,
  startSavingTrace,
  discardTrace,
  uploadPendingTraces
} from '../../app/actions/traces'
const middlewares = [thunk]
const mockStore = configureStore(middlewares)

jest.mock('../../app/services/trace', () => {
  return {
    startTrace: dispatch => {
      dispatch({
        type: 'TRACE_START'
      })
      dispatch({
        type: 'TRACE_SET_SUBSCRIPTION',
        data: {
          remove: jest.fn()
        }
      })
      for (var i = 0; i < 4; i++) {
        dispatch({
          type: 'TRACE_POINT_CAPTURED',
          data: {
            longitude: i,
            latitude: i,
            timestamp: i
          }
        })
      }
    },
    endTrace: (dispatch, watcher, description) => {
      dispatch({
        type: 'TRACE_END',
        description
      })
    }
  }
})

// This is required because the observe API service calls the store directly
// to get the token.
jest.mock('../../app/utils/store', () => {
  return {
    store: {
      getState: () => {
        return {
          observeApi: {
            token: 'abcd'
          }
        }
      }
    }
  }
})

const getMockCurrentTrace = function () {
  return {
    type: 'Feature',
    properties: {
      timestamps: [],
      accuracies: []
    },
    geometry: {
      type: 'LineString',
      coordinates: []
    }
  }
}

/**
 *
 * @param {Number} m - used to construct id, timestamps, coords
 */
const getMockTrace = function (m) {
  return {
    id: `id-${m}`,
    pending: true,
    uploading: false,
    geojson: {
      type: 'Feature',
      properties: {
        timestamps: [
          m,
          m + 10,
          m + 20
        ]
      },
      geometry: {
        type: 'LineString',
        coordinates: [
          [m, m],
          [m + 1, m - 1],
          [m + 2, m - 2]
        ]
      }
    }
  }
}

const getMockTracePostResponse = function (m, id) {
  return {
    type: 'Feature',
    properties: {
      id,
      timestamps: [
        m,
        m + 10,
        m + 20
      ]
    },
    geometry: {
      type: 'LineString',
      coordinates: [
        [m, m],
        [m + 1, m - 1],
        [m + 2, m - 2]
      ]
    }
  }
}

describe('test trace sync actions', () => {
  it('should start trace correctly', () => {
    const store = mockStore({
      traces: {
        currentTrace: null
      }
    })
    store.dispatch(startTrace())
    const actions = store.getActions()
    expect(actions.length).toEqual(6)
    expect(actions).toMatchSnapshot()
  })

  it('should end trace correctly', () => {
    const store = mockStore({
      traces: {
        currentTrace: getMockCurrentTrace()
      }
    })
    store.dispatch(endTrace('test description'))
    const actions = store.getActions()
    expect(actions.length).toEqual(1)
    expect(actions[0].type).toEqual('TRACE_END')
    expect(actions[0].description).toEqual('test description')
  })
  it('should pause trace', () => {
    const store = mockStore({
      traces: {
        currentTrace: getMockCurrentTrace()
      }
    })
    store.dispatch(pauseTrace())
    const actions = store.getActions()
    expect(actions[0].type).toEqual('TRACE_PAUSE')
  })
  it('should unpause trace', () => {
    const store = mockStore({
      traces: {
        currentTrace: getMockCurrentTrace()
      }
    })
    store.dispatch(unpauseTrace())
    const actions = store.getActions()
    expect(actions[0].type).toEqual('TRACE_UNPAUSE')
  })
  it('should start saving trace', () => {
    const store = mockStore({})
    store.dispatch(startSavingTrace())
    const actions = store.getActions()
    expect(actions[0].type).toEqual('TRACE_START_SAVING')
  })
  it('should discard trace', () => {
    const store = mockStore({
      traces: {
        watcher: {
          remove: jest.fn()
        }
      }
    })
    store.dispatch(discardTrace())
    const actions = store.getActions()
    expect(actions[0].type).toEqual('TRACE_DISCARD')
  })
})

describe('trace upload / async actions', () => {
  it('should upload a single trace', async () => {
    const store = mockStore({
      traces: {
        traces: [
          getMockTrace(1)
        ]
      }
    })
    fetch.resetMocks()
    fetch.once(JSON.stringify(getMockTracePostResponse(1, 'fakeid')))
    await store.dispatch(uploadPendingTraces())
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
    expect(fetch.mock.calls).toMatchSnapshot()
  })

  it('should upload multiple pending traces', async () => {
    const store = mockStore({
      traces: {
        traces: [
          getMockTrace(1),
          getMockTrace(2)
        ]
      }
    })
    fetch.resetMocks()
    fetch.once(JSON.stringify(getMockTracePostResponse(1, 'fakeid-1')))
      .once(JSON.stringify(getMockTracePostResponse(2, 'fakeid-2')))
    await store.dispatch(uploadPendingTraces())
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
    expect(fetch.mock.calls).toMatchSnapshot()
  })

  it('should not upload non-pending traces', async () => {
    const trace1 = getMockTrace(1)
    const trace2 = getMockTrace(2)
    trace2.pending = false
    const store = mockStore({
      traces: {
        traces: [
          trace1,
          trace2
        ]
      }
    })
    fetch.resetMocks()
    fetch.once(JSON.stringify(getMockTracePostResponse(1, 'fakeid-1')))
    await store.dispatch(uploadPendingTraces())
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
    expect(fetch.mock.calls).toMatchSnapshot()
  })

  it('should not upload uploading traces', async () => {
    const trace1 = getMockTrace(1)
    trace1.uploading = true
    const trace2 = getMockTrace(2)
    const store = mockStore({
      traces: {
        traces: [
          trace1,
          trace2
        ]
      }
    })
    fetch.resetMocks()
    fetch.once(JSON.stringify(getMockTracePostResponse(2, 'fakeid-2')))
    await store.dispatch(uploadPendingTraces())
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
    expect(fetch.mock.calls).toMatchSnapshot()
  })
})
