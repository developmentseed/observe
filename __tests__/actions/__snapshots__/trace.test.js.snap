// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test trace sync actions should start trace correctly 1`] = `
Array [
  Object {
    "type": "TRACE_START",
  },
  Object {
    "data": Object {
      "remove": [MockFunction],
    },
    "type": "TRACE_SET_SUBSCRIPTION",
  },
  Object {
    "data": Object {
      "latitude": 0,
      "longitude": 0,
      "timestamp": 0,
    },
    "type": "TRACE_POINT_CAPTURED",
  },
  Object {
    "data": Object {
      "latitude": 1,
      "longitude": 1,
      "timestamp": 1,
    },
    "type": "TRACE_POINT_CAPTURED",
  },
  Object {
    "data": Object {
      "latitude": 2,
      "longitude": 2,
      "timestamp": 2,
    },
    "type": "TRACE_POINT_CAPTURED",
  },
  Object {
    "data": Object {
      "latitude": 3,
      "longitude": 3,
      "timestamp": 3,
    },
    "type": "TRACE_POINT_CAPTURED",
  },
]
`;

exports[`trace upload / async actions should not upload non-pending traces 1`] = `
Array [
  Object {
    "id": "id-1",
    "type": "TRACE_UPLOAD_STARTED",
  },
  Object {
    "newId": "fakeid-1",
    "oldId": "id-1",
    "type": "TRACE_UPLOADED",
  },
]
`;

exports[`trace upload / async actions should not upload non-pending traces 2`] = `
Array [
  Array [
    "http://localhost:3000/traces",
    Object {
      "body": "{\\"tracejson\\":{\\"type\\":\\"Feature\\",\\"properties\\":{\\"timestamps\\":[1,11,21],\\"description\\":\\"\\"},\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[1,1],[2,0],[3,-1]]}}}",
      "headers": Object {
        "Authorization": "abcd",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`trace upload / async actions should not upload uploading traces 1`] = `
Array [
  Object {
    "id": "id-2",
    "type": "TRACE_UPLOAD_STARTED",
  },
  Object {
    "newId": "fakeid-2",
    "oldId": "id-2",
    "type": "TRACE_UPLOADED",
  },
]
`;

exports[`trace upload / async actions should not upload uploading traces 2`] = `
Array [
  Array [
    "http://localhost:3000/traces",
    Object {
      "body": "{\\"tracejson\\":{\\"type\\":\\"Feature\\",\\"properties\\":{\\"timestamps\\":[2,12,22],\\"description\\":\\"\\"},\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[2,2],[3,1],[4,0]]}}}",
      "headers": Object {
        "Authorization": "abcd",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`trace upload / async actions should upload a single trace 1`] = `
Array [
  Object {
    "id": "id-1",
    "type": "TRACE_UPLOAD_STARTED",
  },
  Object {
    "newId": "fakeid",
    "oldId": "id-1",
    "type": "TRACE_UPLOADED",
  },
]
`;

exports[`trace upload / async actions should upload a single trace 2`] = `
Array [
  Array [
    "http://localhost:3000/traces",
    Object {
      "body": "{\\"tracejson\\":{\\"type\\":\\"Feature\\",\\"properties\\":{\\"timestamps\\":[1,11,21],\\"description\\":\\"\\"},\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[1,1],[2,0],[3,-1]]}}}",
      "headers": Object {
        "Authorization": "abcd",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`trace upload / async actions should upload multiple pending traces 1`] = `
Array [
  Object {
    "id": "id-1",
    "type": "TRACE_UPLOAD_STARTED",
  },
  Object {
    "newId": "fakeid-1",
    "oldId": "id-1",
    "type": "TRACE_UPLOADED",
  },
  Object {
    "id": "id-2",
    "type": "TRACE_UPLOAD_STARTED",
  },
  Object {
    "newId": "fakeid-2",
    "oldId": "id-2",
    "type": "TRACE_UPLOADED",
  },
]
`;

exports[`trace upload / async actions should upload multiple pending traces 2`] = `
Array [
  Array [
    "http://localhost:3000/traces",
    Object {
      "body": "{\\"tracejson\\":{\\"type\\":\\"Feature\\",\\"properties\\":{\\"timestamps\\":[1,11,21],\\"description\\":\\"\\"},\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[1,1],[2,0],[3,-1]]}}}",
      "headers": Object {
        "Authorization": "abcd",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "http://localhost:3000/traces",
    Object {
      "body": "{\\"tracejson\\":{\\"type\\":\\"Feature\\",\\"properties\\":{\\"timestamps\\":[2,12,22],\\"description\\":\\"\\"},\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[2,2],[3,1],[4,0]]}}}",
      "headers": Object {
        "Authorization": "abcd",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;
